driver/Src/CAN.c

```diff
@@ -0,0 +1,377 @@
+#include "stm32xx_hal.h"
+#include "CAN.h"
+
+// 8 for now unless extended payload is supported
+#define DATA_SIZE (8)
+
+// entries in queues
+typedef struct {
+    CAN_TxHeaderTypeDef header;
+    uint8_t data[DATA_SIZE];
+} tx_payload_t;
+
+typedef struct {
+    CAN_RxHeaderTypeDef header;
+    uint8_t data[DATA_SIZE];
+} rx_payload_t;
+
+// metadata for recieve queues
+typedef struct {
+    uint16_t id;
+    uint16_t size;
+    QueueHandle_t queue;
+    uint8_t* storage;
+    StaticQueue_t buffer;
+} recv_entry_t;
+
+#ifdef CAN1
+// fallback send queue size
+#ifndef CAN1_SEND_QUEUE_SIZE
+#define CAN1_SEND_QUEUE_SIZE (10)
+#endif
+
+// can1 handle
+static CAN_HandleTypeDef hcan1_ = {.Instance = CAN1};
+CAN_HandleTypeDef* hcan1 = &hcan1_;
+
+// can1 send semaphore
+static SemaphoreHandle_t can1_send_semaphore = NULL;
+static StaticSemaphore_t can1_send_semaphore_buffer;
+
+// can1 send queue
+static QueueHandle_t can1_send_queue = NULL;
+static StaticQueue_t can1_send_queue_buffer;
+static uint8_t can1_send_queue_storage[CAN1_SEND_QUEUE_SIZE * sizeof(tx_payload_t)];
+
+// can1 recv semaphore
+static SemaphoreHandle_t can1_recv_semaphore = NULL;
+static StaticSemaphore_t can1_recv_semaphore_buffer;
+
+// create recv queue storage
+#define CAN_RECV_ENTRY(ID_, SIZE_) \
+    static uint8_t recv_queue_storage_##ID_[SIZE_ * sizeof(rx_payload_t)]; \
+
+#include "can_entries.h"
+
+#undef CAN_RECV_ENTRY
+
+// create recv queue array
+#define CAN_RECV_ENTRY(ID_, SIZE_) \
+    { \
+        .id = (ID_), \
+        .size = (SIZE_), \
+        .queue = NULL, \
+        .storage = recv_queue_storage_##ID_, \
+        .buffer = {{0}} \
+    },
+
+static recv_entry_t can1_recv_entries[] = {
+    #include "can_entries.h"
+};
+#undef CAN_RECV_ENTRY
```

Comment:
Or is the user expected to define a can_entries.h file which includes the whitelisted entries? If so, could this not just be a big struct like we had last year with all the config information rather than this whole macro thing

Reply:
Also, yeah. I didn't think about how it works with the whitelist of the filterconfig. To be honest, I can just make the filter list at run time based off this struct so the programmer doesn't even need to make the filter config either

Reply:
I thought about making it a big struct but the issue is if we want the queue sizes to be different. Which maybe could be the case if one can entry is very chatty and we want more space for it compared to the others

Reply:
I suppose that makes sense. Although, if the queue sizes are different, could the buffer just be stored separately and pointed to by the struct? That would fix the size problem so now every entry is same size

Reply:
yes but then someone needs to make the queue storage right? Either I do that with macros or I let the programmer do it. Why would I let the programmer do it if I can do it without error with macros? Better for the programmer to not worry about making the storage if they don't need to

Reply:
That makes sense I spose. Would we rather have this as a more shared config rather than a CAN-specific config, as other protocols will also need this sort of queue creation behavior? At least for something like I2C or SPI or smth.

Reply:
@NathanielDelgado